okhttp/src/main/java/okhttp3/internal/connection/RealConnection.java

```diff
@@ -376,8 +387,78 @@ private Request createTunnelRequest() {
   /** Returns true if this connection can carry a stream allocation to {@code address}. */
   public boolean isEligible(Address address) {
     return allocations.size() < allocationLimit
-        && address.equals(route().address())
-        && !noNewStreams;
+        && !noNewStreams
+        && canCarryAddress(address);
+  }
+
+  private boolean canCarryAddress(Address address) {
+    boolean directMatch = address.equals(route().address());
+
+    if (directMatch) {
+      return true;
+    }
+
+    if (address.isCoalescable() && this.isCoalescable() && address.equalsNonHost(route.address())) {
```

Comment:
As I read it, there’s four requirements here:

 * address.isCoalescable
 * this.isCoalescable
 * address.equalsNonHost(route.address())
 * supportsHost()

Should this just be one expression?